Monitoring of Raft message processing performance

andrewkerr9000 opened the pull request:
Delay is difference between unmarshalling of message and start of
processing.

andrewkerr9000 committed:
Refactor LeaderAvailabilityHandler to visitor pattern

andrewkerr9000 committed:
Monitoring for delay in Raft message processing performance

Delay is duration between unmarshalling of message and start of
processing.

Timing is the duration to process each raft message, collectively
and grouped by message type.

andrewkerr9000 commented:
Diff is long but simple.

I experimented with a message wrapper (similar to `ClusterIdAwareMessage`) to include the `receivedAt` Instant, but that was looking uglier.

Passing `Clock` instances around is not strictly necessary.

andrewkerr9000 commented:
There may be some room for improvement with timer, e.g. a different named timer for each message type.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java:
```diff
@@ -1017,13 +1032,17 @@ public int hashCode()
-        private Type type;
+        protected final MemberId from;
+        private final Type type;
+        private final Clock clock;
```
I might be able to live with the recievedAt, but not the clock. Perhaps you can add another handler in the Netty-pipeline which fills in recievedAt?

andrewkerr9000 replied:
Keeping a reference to the clock makes it easier to reuse for creating messages in response. Alternatives aren't great:

* Don't inject a Clock, hard code instead
* Allow a null Instant, or make Optional
* Modify message handle methods to take a clock
* Make role handlers not singletons and inject clock
* Wrap messages in a timing aware wrapper similar to ClusterIdAwareMessage. Tried this, all of plain message / cluster ID message / timing message / cluster ID and timing message occur, it got messy. Could be the best solution, would be nicer in Scala :smiling_imp:

martinfurmanski replied:
I would be in favour of "Allow a null Instant, or make Optional" because this field is not really part of the message proper and should not be serialized. It is merely an additional tag on the message that gets added when it passes through the processing layers. Wrapping it would also be ok, but just having a special field is perhaps easier to work with and is also low overhead.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMachine.java:
```diff
@@ -280,6 +308,24 @@ public synchronized ConsensusOutcome handle( RaftMessages.RaftMessage incomingMe
+        long electionTimeout = leaderAvailabilityTimers.getElectionTimeout();
+        if ( delay >= electionTimeout )
+        {
+            log.warn( "Message %s processing delay %s greater than election timeout interval %s", message, delay, electionTimeout );
```
I am skeptical to this warning, which could have a great spamming and unnecessary support-inducing potential. Having the metrics available should perhaps be a sufficient solution?

martinfurmanski reviewed (changes requested):
I am slightly skeptical to this overall because it sort of messes with the layering of code and measures something very deep in the stack. The worst offender here though with the clock I believe can be fixed.

I think it has potential to be useful for our internal debugging of the system going forward though, and perhaps even generally for customers in a little longer perspective.

I would like to see this in action in one of our more stressful workloads.

I am also missing tests showing that the various timers work and also Java docs describing them.

andrewkerr9000 commented:
I am a little concerned that the monitoring, in particular the Timers, could appreciably impact performance or memory footprint.

There is (currently) one Timer per message type, plus one for all types. There could be one per role / message type combination, which may be excessive.

andrewkerr9000 commented:
I'm thinking of inserting a new MessageHandler in front of the RaftMachine to be responsible for monitoring RaftMachine.

It may then be worth introducing a class to be responsible for pipelining message handlers to make it clearer how they are wired up.

andrewkerr9000 commented:
Added a received-at aware wrapper for Raft messages and a monitoring-dedicated handler. There are a couple of further refactorings I want to look at too.

andrewkerr9000 committed:
Refactor CoreServerModule, message handlers

Introduce RaftServerModule, chain message handlers using functional
composition

Delegate lifecycle methods along message handler chain

Extract cluster binding check to separate message handler class

Rename RaftMessageHandler class to RaftMessageApplier
Should make clearer that it is responsible for executing outcome of
application of raft machine with raft message.

andrewkerr9000 commented:
Introduced a module for creating chained raft message handler, with dependencies. Refactored handler chaining to be functional composition. Refactored life cycle handling so that head of message handler chain is added to core life, and delegates onwards - this now includes the job scheduler for BatchingMessageHandler. Also split out cluster binding logic into a dedicated handler, with two instances in handler chain - the second in on the processing thread but may be redundant.

Will squash.

Will also fire up soak tests, both to observe monitoring and to check I haven't broken anything - PR is quite invasive now.

May also improve unit test coverage on RaftMessageHandler/RaftMessageApplier

andrewkerr9000 committed:
Refactor enriched raft message hierarchy, BatchingMessageHandler

andrewkerr9000 commented:
@martinfurmanski just about finished messing about with this now. Care to take another look? Hopefully clear to review commit by commit.

martinfurmanski commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/core/consensus/CatchUpTest.java:
```diff
@@ -95,9 +95,9 @@ public void newMemberWithNoLogShouldCatchUpFromPeers() throws Throwable
         net.disconnect( sleepyId );
 
         // when
+        fixture.members().withId( leaderId ).raftInstance().handle( new Request( leader, valueOf( 30 ) ) );
```
Was this an intended move of this line?

andrewkerr9000 replied:
Doubtful. Will revert.

martinfurmanski commented on enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/core/consensus/roles/PreVotingVoterTest.java:
```diff
@@ -52,7 +52,8 @@ public void shouldRespondPositivelyIfWouldVoteForCandidate() throws Exception
 
         // when
-        Outcome outcome = new Follower().handle( new RaftMessages.PreVote.Request( member1, 0, member1, 0, 0 ), raftState, log() );
+        Outcome outcome = new Follower().handle( new RaftMessages.PreVote.Request( member1, 0, member1, 0, 0 ),
```
This file seems unnecessarily reformatted.

andrewkerr9000 replied:
Well spotted. This would have been a left over from the abandoned approach of including an Instant in all RaftMessages.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/RaftServerModule.java:
```diff
@@ -0,0 +1,117 @@
+        return clusterBindingHandler
+                .compose( leaderAvailabilityHandler )
+                .compose( batchingMessageHandler )
+                .compose( clusterBindingHandler )
```
Should the cluster-binding really be twice in the stack here?

andrewkerr9000 replied:
It's the existing logic. I expect it's unnecessary.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java:
```diff
@@ -979,38 +1006,14 @@ public String toString()
     }
 
-    class ReceivedInstantAwareMessageImpl implements ReceivedInstantAwareMessage
+    class ReceivedInstantAwareMessageImpl<RM extends RaftMessage> implements ReceivedInstantAwareMessage<RM>
```
I wonder if it would be possible and beneficial to flatten the type-hierarchy of raft messages, to just have a single base RaftMessage type with all the different behaviours in it; ReceivedInstant, default implementations of dispatch, ClusterId, and so on.

andrewkerr9000 replied:
Not all RaftMessage instances have (or need) a ReceivedInstant, or a ClusterId, or either. This way keeps all the complexity in one place.

martinfurmanski reviewed (approved):
Minor comments. Checking approve, but fix what makes sense.

andrewkerr9000 committed:
Remove redundant cluster binding raft message handler
